### PR TITLE
AWS S3 Bindings Certification Tests

### PR DESCRIPTION
**Pull request by robertojrojas**
# Description

Provides the AWS S3 Bindings Conformance and Certification Test Plan and Test Code.

## Issue reference

Please reference the issue this PR will close: #2247

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests

### CODE REVIEW

File: .github/scripts/test-info.mjs
Comment (berndverst) [inline .github/scripts/test-info.mjs:85]: Something seems off here.

For the Docker (localstack) case it should say `certification: false` and also `requireAWSCredentials: false` and `requireTerraform: false`

You you should only run local stack for conformance test.

Of course you will also run the terraform conformance test with regular AWS. Certification is AWS + Terraform only.

In general I highly recommend you refactor `bindings.aws.s3` to `bindings.aws.s3.localstack` to make it clear what is what. I have seen strange issues with nested tests executing all the tests. It is best if these are in sibling folders.

**Review by berndverst — CHANGES_REQUESTED** [review]

File: .github/scripts/test-info.mjs
Comment (berndverst) [inline .github/scripts/test-info.mjs:85]: Let's remove this entry as discussed? We don't want to run the tests with docker in conformance and certification test.

Comment (robertojrojas) [reply .github/scripts/test-info.mjs:85]: @berndverst absolutely! I'll take care of that today.

File: .github/scripts/test-info.mjs
Comment (ItalyPaleAle) [inline .github/scripts/test-info.mjs:560]: I think it's ok to run Docker and Cloud  tests together. The two things aren't mutually exclusive.

Comment (ItalyPaleAle) [reply .github/scripts/test-info.mjs:560]: After speaking with @robertojrojas i see the benefit of having 2 separate tests

File: tests/certification/bindings/aws/s3/go.mod
Comment (berndverst) [inline tests/certification/bindings/aws/s3/go.mod:3]:
```suggestion
go 1.20
```

**Review by berndverst — CHANGES_REQUESTED** [review]
Can you please use `encodeBase64` in one of your certification tests and verify the behavior?

**Comment by berndverst** [comment]
@robertojrojas seems our AWS service account doesn't have the necessary permissions. Perhaps @artursouza can help you get that taken care of.

**Comment by robertojrojas** [comment]
@berndverst @ItalyPaleAle thanks for reviewing this PR. I've added a cert test for `encodeBase64` and I'll coordinate with @artursouza regarding the AWS S3 permissions.

File: tests/certification/bindings/aws/s3/s3_test.go
Comment (artursouza) [inline tests/certification/bindings/aws/s3/s3_test.go:84]: nit: typo

File: tests/certification/bindings/aws/s3/s3_test.go
Comment (artursouza) [inline tests/certification/bindings/aws/s3/s3_test.go:146]: What is the difference between these tests and the existing conformance tests?

File: tests/conformance/bindings/bindings.go
Comment (artursouza) [inline tests/conformance/bindings/bindings.go:240]: Is this a legit behavior change? Can resp be nil even when create was performed?

Usually, changes in conformance tests need to be carefully reviewed to understand if the variation of the component is acceptable.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: tests/certification/bindings/aws/s3/s3_test.go
Comment (robertojrojas) [reply tests/certification/bindings/aws/s3/s3_test.go:146]: One difference I see is the fact the conformance tests require either `docker` or `cloud` resources to be ready prior testing.

File: tests/conformance/bindings/bindings.go
Comment (robertojrojas) [reply tests/conformance/bindings/bindings.go]: I rememeber see the `resp` with nil, but I can't reproduce the issue any longer, so I'm removing that change.

File: tests/certification/bindings/aws/s3/s3_test.go
Comment (artursouza) [reply tests/certification/bindings/aws/s3/s3_test.go:146]: Is there any different in the test scenarios?

Comment (robertojrojas) [reply tests/certification/bindings/aws/s3/s3_test.go:146]: For S3 Cert test in particular, it will test certain metadata tags that are not necessarily  common with other components like **forcePathStyle**, **encodeBase64**,  and **decodeBase64**.  I hope that answers your question.

Comment (artursouza) [reply tests/certification/bindings/aws/s3/s3_test.go:146]: It does. Thanks.

**Review by artursouza — DISMISSED** [review]
Please, wait for Bernd to approve or dismiss his review.

**Review by berndverst — APPROVED** [review]

**Pull request merged**
